### PR TITLE
(REPLATS-457) Handle ingress for KinD clusters

### PR DESCRIPTION
**Pull request by jpartlow**
* Handles KinD specific ingress check and start ingress needs
* Makes pam_tools::delete_kots_app idempotent if kotsadm not present
* Makes pam_tools::delete_kotsadm skip Kurl hosts unless forced
* Fixes some README toc markdown

You still get an error if you re-run delete_kots_app for a non-existent slug when kotsadm is still running, but I'm not going to worry about that right now.

### CODE REVIEW

File: tasks/has_ingress_controller.rb [hunk 1]
```diff
@@ -8,15 +8,16 @@ def lb_port?(spec, port)
...
 
-  def node_port?(spec, port)
-    spec['type'] == 'NodePort' && spec['ports'].any? { |p| p['nodePort'] == port }
+  def node_port?(spec, port, port_param)
+    spec['type'] == 'NodePort' && spec['ports'].any? { |p| p[port_param] == port }
   end
 
   def task(**_kwargs)
     services = get_all_services
+    node_port_parameter = kind_cluster? ? 'port' : 'nodePort'
```
[the remark this answers is not in the document]
Comment (jpartlow) [reply tasks/has_ingress_controller.rb:17]: Sorry, I've be staring at container images and forgot about this. But here it is, I think:

On the engineering-scratchpad Kind host jpartlow-kind-ubuntu, which is 10.253.0.199, the k8s container's ip address is 172.18.0.2:

```
jpartlow@jpartlow-kind-ubuntu:~$ docker ps
CONTAINER ID   IMAGE                  COMMAND                  CREATED        STATUS      PORTS                                                                 NAMES
8edb87868483   kindest/node:v1.20.2   "/usr/local/bin/entr…"   2 months ago   Up 8 days   0.0.0.0:80->80/tcp, 0.0.0.0:443->443/tcp, 127.0.0.1:45361->6443/tcp   kind-control-plane
jpartlow@jpartlow-kind-ubuntu:~$ docker inspect 8edb87868483 | grep '"IPAddress":'
            "IPAddress": "",
                    "IPAddress": "172.18.0.2",
```

And docker-proxy is forwarding ports 80, 443 from any incoming to that address:

```
jpartlow@jpartlow-kind-ubuntu:~$ ps -ef | grep docker-proxy
root        2391    2156  0 Oct25 ?        00:00:30 /usr/bin/docker-proxy -proto tcp -host-ip 127.0.0.1 -host-port 45361 -container-ip 172.18.0.2 -container-port 6443
root        2407    2156  0 Oct25 ?        00:00:00 /usr/bin/docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 443 -container-ip 172.18.0.2 -container-port 443
root        2421    2156  0 Oct25 ?        00:00:00 /usr/bin/docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 80 -container-ip 172.18.0.2 -container-port 80
```

Internally in the container, no process is listening on 80 or 443, but if you look at the container's iptables nat table with `docker exec -it kind-control-plane /bin/bash -c 'iptables -t nat -L -v -n'`, you see these entries dnatting 80/443 to 10.244.0.23 which is the ingress-nginx-controller pod's ip:

```
Chain PREROUTING (policy ACCEPT 175 packets, 10500 bytes)
 pkts bytes target     prot opt in     out     source               destination
4395K  334M KUBE-SERVICES  all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* kubernetes service portals */
  562 46670 DOCKER_OUTPUT  all  --  *      *       0.0.0.0/0            172.18.0.1
 6473  388K CNI-HOSTPORT-DNAT  all  --  *      *       0.0.0.0/0            0.0.0.0/0            ADDRTYPE match dst-type LOCAL

...

Chain CNI-DN-a957e0fb5e37e93333b7c (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 CNI-HOSTPORT-SETMARK  tcp  --  *      *       10.244.0.0/24        0.0.0.0/0            tcp dpt:80
    0     0 CNI-HOSTPORT-SETMARK  tcp  --  *      *       127.0.0.1            0.0.0.0/0            tcp dpt:80
    6   360 DNAT       tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:80 to:10.244.0.23:80
    0     0 CNI-HOSTPORT-SETMARK  tcp  --  *      *       10.244.0.0/24        0.0.0.0/0            tcp dpt:443
    0     0 CNI-HOSTPORT-SETMARK  tcp  --  *      *       127.0.0.1            0.0.0.0/0            tcp dpt:443
    7   420 DNAT       tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:443 to:10.244.0.23:443

Chain CNI-HOSTPORT-DNAT (2 references)
 pkts bytes target     prot opt in     out     source               destination
   13   780 CNI-DN-a957e0fb5e37e93333b7c  tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            /* dnat name: "kindnet" id: "85070a5e0330c619ebc45ff1cc9c20337e640f078537c6a5a0006561c1452611" */ multiport dports 80,443
```

So, the pod hostPort plus docker-proxy are ensuring that the vm's internal ip 80/443 make their way to the ingress, and in this case, the service nodePort isn't really involved (there are separate dnat's that ensure that container traffic on those ports make their way to the ingress but nothing is connecting those nodePort's to the outside that I can see).

[the remark this answers is not in the document]
Comment (MikaelSmith) [reply tasks/has_ingress_controller.rb:17]: `hostPort` is a common way to configure an ingress gateway.

[the remark this answers is not in the document]
Comment (MikaelSmith) [reply tasks/has_ingress_controller.rb:17]: `hostPort` is a common way to configure an ingress gateway.

[the remark this answers is not in the document]
Comment (jpartlow) [reply tasks/has_ingress_controller.rb:17]: Alright, I'll add a test for any pod with an http hostPort instead.

[the remark this answers is not in the document]
Comment (jpartlow) [reply tasks/has_ingress_controller.rb]: That's in.

**Review by MikaelSmith — APPROVED** [review]

**Pull request merged**